src/coins.cpp

```diff
@@ -67,15 +72,24 @@ size_t CCoinsViewCache::ResetCachedCoinUsage() const
     return newCachedCoinsUsage;
 }
 
-CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
+CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint, CDeferredSharedLocker *lock) const
 {
-    AssertLockHeld(cs_utxo);
-    CCoinsMap::iterator it = cacheCoins.find(outpoint);
-    if (it != cacheCoins.end())
-        return it;
+    // AssertLockHeld(cs_utxo);
+    {
+        if (lock)
+            lock->lock_shared();
+        CCoinsMap::iterator it = cacheCoins.find(outpoint);
+        if (it != cacheCoins.end())
+            return it;
+        if (lock)
+            lock->unlock();
+    }
     Coin tmp;
     if (!base->GetCoin(outpoint, tmp))
         return cacheCoins.end();
+
+    if (lock)
```

Comment:
I think a few comments in here about the need for  first a shared lock above and then an exlusive lock below would be helpful for anyone unfamiliar with this section.  All the lock and lock->lock() lock->shared() make it difficult to pick out, at least for me.  I was scratching my head for a while and wondering by you were'nt using an exclusive lock in before updating the map when in fact you were - lol.

Reply:
yes this section was conceptually quite tricky.  Basically, the problem is that even though the caller may be just reading data, we may update the cache and in that case we need the write lock.  I will add a comment.